### PR TITLE
Fixes RequestManager handling and adding of the 'data' listener

### LINKED ISSUE #1648: MaxListenersExceededWarning when sending contract (v1.0.0-beta.33)
**Situation**: When I send a deployed contract using `contract.deploy().send()`, the command line prints a warning:
```
  contracts
(node:23488) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 data listeners added. Use emitter.setMaxListeners() to increase limit
    √ gets the accounts

  6 passing (1s)
```

_Edit (14 April 2020): removed unnecessary information_

### PR DESCRIPTION
**Pull request by nivida**
Fixes #1648, #3042 

### Descriptions

Fixes the ``MaxEventListener`` issue with adding of the ``setRequestManager`` function to all packages. The provider will now only get resolved once and the ``data`` listener is no longer added on each initiation of a ``Module``, ``Contract``, or on executing of the ``setProvider`` method on each package.

**Old Behaviour**

1. call ``setProvider``.
2. create RequestManager and register ``data`` listener.
3. call ``setProvider`` on all children's.
4. create RequestManager and register ``data`` listener for each child package.

**New Behaviour**

1. call ``setProvider``.
2. create RequestManager and register ``data`` listener.
3. call ``setRequestManager`` on all children's.

### CODE REVIEW

**Review by cgewecke — CHANGES_REQUESTED** [review]
Fixes the warning which is awesome. 💯 💯 

+ [without this PR](https://travis-ci.org/ethereum/web3.js/jobs/603601765#L849-L851)
+ [with this PR](https://travis-ci.org/ethereum/web3.js/jobs/605095785#L878-L879)

Is there anyway to get the diff into a state where only the meaningful logic changes are represented before going over for a quick second pass?

**Comment by nivida** [comment]
@cgewecke Because those code style improvements are made within commits with actually fixes is it really hard to remove those changes. I will not do such changes to newly proposed PRs. Thanks for the hint!

File: packages/web3-eth-contract/src/index.js [hunk 1]
```diff
@@ -56,22 +56,18 @@ var Contract = function Contract(jsonInterface, address, options) {
         args = Array.prototype.slice.call(arguments);
 
     if(!(this instanceof Contract)) {
-        throw new Error('Please use the "new" keyword to instantiate a web3.eth.contract() object!');
+        throw new Error('Please use the "new" keyword to instantiate a web3.eth.Contract() object!');
     }
 
     // sets _requestmanager
-    core.packageInit(this, [this.constructor.currentProvider]);
+    core.packageInit(this, [this.constructor]);
```
Comment (cgewecke) [inline packages/web3-eth-contract/src/index.js:63]: Flagging as logic change.

File: packages/web3-eth-contract/src/index.js [hunk 2]
```diff
@@ -606,18 +613,20 @@ Contract.prototype.once = function(event, options, callback) {
...
  * @method _on
+ *
  * @param {String} event
  * @param {Object} options
  * @param {Function} callback
+ *
  * @return {Object} the event subscription
  */
-Contract.prototype._on = function(){
+Contract.prototype._on = function(event, options, callback){
```
Comment (cgewecke) [inline packages/web3-eth-contract/src/index.js]: Logic change

File: packages/web3-eth-contract/src/index.js [hunk 3]
```diff
@@ -606,18 +613,20 @@ Contract.prototype.once = function(event, options, callback) {
...
     var subOptions = this._generateEventOptions.apply(this, arguments);
 
 
     // prevent the event "newListener" and "removeListener" from being overwritten
-    this._checkListener('newListener', subOptions.event.name, subOptions.callback);
-    this._checkListener('removeListener', subOptions.event.name, subOptions.callback);
+    this._checkListener('newListener', subOptions.event.name);
+    this._checkListener('removeListener', subOptions.event.name);
```
Comment (cgewecke) [inline packages/web3-eth-contract/src/index.js:629]: Logic change

File: packages/web3-eth-ens/src/contracts/Registry.js [hunk 4]
```diff
@@ -36,13 +35,11 @@ var RESOLVER_ABI = require('../ressources/ABI/Resolver');
...
-    var self = this;
+    const self = this;
     this.ens = ens;
     this.contract = ens.checkNetwork().then(function (address) {
-        var contract = new Contract(REGISTRY_ABI, address);
-        contract.setProvider(self.ens.eth.currentProvider);
-
-        return contract;
+        return new self.ens.eth.Contract(REGISTRY_ABI, address);
+        // contract.setProvider(self.ens.eth.currentProvider);
```
Comment (cgewecke) [inline packages/web3-eth-ens/src/contracts/Registry.js]: Logic change.

@nivida Do you want to preserve this commented out line?

File: packages/web3-eth-contract/src/index.js (hunk 3 above)
Comment (nivida) [reply packages/web3-eth-contract/src/index.js:629]: The passed parameter doesn't exist on the ``_checkListener`` method and isn't a logic change.

File: packages/web3-eth-ens/src/contracts/Registry.js (hunk 4 above)
Comment (nivida) [reply packages/web3-eth-ens/src/contracts/Registry.js]: Thanks for seeing this! I will remove this line 👌

File: packages/web3-core/src/index.js [hunk 5]
```diff
@@ -49,36 +48,40 @@ module.exports = {
...
                 pkg._requestManager.setProvider(provider, net);
                 pkg._provider = pkg._requestManager.provider;
                 return true;
             };
         }
 
+        pkg.setRequestManager = function(manager) {
+            pkg._requestManager = manager;
+            pkg._provider = manager.provider;
+        };
```
Comment (cgewecke) [inline packages/web3-core/src/index.js:76]: Logic change

Comment (nivida) [reply packages/web3-core/src/index.js:76]: This is a required logic change and also described in the description of this PR:
> Fixes the MaxEventListener issue with adding of the setRequestManager function to all packages.

Comment (cgewecke) [reply packages/web3-core/src/index.js:76]: I'm flagging these so I can see what's material - that's all - sorry.

File: packages/web3-eth-contract/src/index.js (hunk 3 above)
Comment (cgewecke) [reply packages/web3-eth-contract/src/index.js:629]: Ah so this was always wrong?

File: packages/web3-eth-contract/src/index.js (hunk 1 above)
Comment (nivida) [reply packages/web3-eth-contract/src/index.js:63]: This is part of the fix and is required to have the inheritance of the web3 modules working.

**Review by cgewecke — COMMENTED** [review]
This will take some time to review, it touches some of the most sensitive logic at Web3. The original design seems a little unusual. I think we need to go through the git blame and make sure we understand why it's currently like this.

@nivida Did you ever discuss the RequestManager logic with Frozeman? Do you remember what his views about it were?

**Comment by cgewecke** [comment]
@frozeman Just pinging you here in case you might remember running into any specific difficulties or design questions around how the RequestManager and setProvider are hooked together.

Do you have any ideas for things to test / double-check?

**Comment by frozeman** [comment]
Without reading all the comments. One reason why I made it this way, was so that certain modules can have a storage privet from the global one. 

Eg. Swarm or whisper can (did) run on separate binaries, and Providers.

File: packages/web3-eth-contract/src/index.js (hunk 3 above)
Comment (nivida) [reply packages/web3-eth-contract/src/index.js:629]: Yep, it sadly was.

**Comment by nivida** [comment]
> This will take some time to review, it touches some of the most sensitive logic at Web3. The original design seems a little unusual. I think we need to go through the git blame and make sure we understand why it's currently like this.

I've went through anything and wrote the logic shortly down on paper to be sure I'm not breaking anything during fixing it. 

> @nivida Did you ever discuss the RequestManager logic with Frozeman? Do you remember what his views about it were?

I've discussed this logic around a year ago and the rule was that child modules do inherit the providers from their parents. This means that providers and default options a module can have do overwrite the values of the child modules when a new value/provider got set but not vice versa. 

@frozeman 
> Without reading all the comments. One reason why I made it this way, was so that certain modules can have a storage privet from the global one.
Eg. Swarm or whisper can (did) run on separate binaries, and Providers.

Yep, this is the implemented logic. 👌 


@cgewecke 
The big difference is that it does only resolve the provider once now and does set it downstream on all child modules without resolving the provider again and without creating a new RequestManager instance for each module and Contract. Because it was resolving the provider and creating the RequestManager object all the time again was it register the ``data`` listener N times.

**Comment by cgewecke** [comment]
@nivida 

> Because it was resolving the provider and creating the RequestManager object all the time again was it register the data listener N times.

Yes! I understand this - it's good. I would still like to take a block of time to look through this a bit to think about it, sorry...

**Comment by nivida** [comment]
Moved this fix to #3190

**Pull request closed without merging**